### DISCUSSION TITLE
Stopping Firefox Browser Handler From Opening Too Many Tabs

### DISCUSSION THREAD

**Discussion by vlombardi22** [cmu-sei/GHOSTS#153] [Q&A] [answered]
Is there a way to stop Firefox agents from opening multiple tabs. Some of the agents I am using are on vm's that do not have a ton of memory so I would like to restrict them to one tab for the browsing timeline. I am just using the Firefox sample timeline with a shorter delay after time and a different set of urls.

**Comment by sei-dupdyke** [comment, marked as answer]
In general, each timeline handler equals an instance of that application. So, there should be just 1 tab per handler that you setup in the default timeline, but there are some scenarios where another tab might be opened periodically:

- if geckodriver, or geckodriver's connection to the browser somehow is lost, the current tab is then basically hung, and ghosts opens a new instance of Firefox and now you have two tabs — the hung one and the new fresh connection. Ghosts periodically cleans up applications and should reap the hung tab after some time.
- if you have other timelines in play, those Firefox handlers would open a new tab for each handler.

Some options that affect resource use that you might consider are:

```
{
   "HandlerType": "BrowserFirefox",
   "HandlerArgs": {
         "isheadless": "true",
         "blockimages": "true",
         "blockstyles": "true",
         "blockflash": "true",
         "blockscripts": "true",
```

Of these, “isheadless” seems to really cut resource use since there is no GUI with that option enabled.

HTH

**Reply by vlombardi22** [reply]
I will give it a shot. How often does ghosts reap the hung tabs?

**Comment by sei-dupdyke** [comment]
It should be ~5 minutes.
